File: 112-path-sum/path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool hasPathSum(TreeNode* root, int targetSum) {
        if(root ==NULL){
            return false;
        }

        if(root->left == NULL && root->right == NULL){
            if(root->val == targetSum){
                return true;
            }
        }

        auto lef = hasPathSum(root->left,targetSum-root->val);

        if(lef){
            return true;
        }
        auto righ = hasPathSum(root->right,targetSum-root->val);

        if(righ){
            return true;
        }

        return false;
    }
};

File: 112-path-sum/path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    boolean ans = false;
    public void checkPathSum(TreeNode root,int targetSum){
        if(root==null){
            return;
        }

        if(root.left == null && root.right==null){
            if(targetSum-root.val ==0){
                ans = true;
            }
            return;
        }

        checkPathSum(root.left,targetSum-root.val);
        checkPathSum(root.right,targetSum-root.val);
    }
    public boolean hasPathSum(TreeNode root, int targetSum) {
       /*
       What should I have to do? 
       -> Root -> leaf   ===> targetSum honi chahiye.        
        */
        // int[] arr = new int[5];


        checkPathSum(root,targetSum);
        return ans;
    }
}